### PR TITLE
Implement KIP-345 in aiokafka (rebase of #682) 

### LINKED ISSUE #680: Support for static membership protocol
**Describe the solution you'd like**
Implement support for static group membership to avoid unnecessary rebalances
[KIP-345](https://cwiki.apache.org/confluence/display/KAFKA/KIP-345%3A+Introduce+static+membership+protocol+to+reduce+consumer+rebalances)

**Additional context**
It looks like this feature is implemented in a fork of this library
https://github.com/robinhood/aiokafka/pull/21

### PR DESCRIPTION
**Pull request by ntextreme3**
### Changes

Fixes #680 

This is a rebase of https://github.com/aio-libs/aiokafka/pull/682 -- see https://github.com/aio-libs/aiokafka/pull/682#issuecomment-1070337629

### Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by ntextreme3** [comment]
@ods Do you have suggestions on the types of tests you'd like to see here. I haven't played with this yet, just rebased it. But I'll need this functionality in the coming months so at some point I'll be going through it more thoroughly anyway.

File: aiokafka/client.py [hunk 1]
```diff
@@ -587,6 +588,7 @@ def _check_api_version_response(self, response):
             # ((2, 6, 0), DescribeClientQuotasRequest[0]),
             ((2, 5, 0), DescribeAclsRequest_v2),
             ((2, 4, 0), ProduceRequest[8]),
+            ((2, 3, 0), SyncGroupRequest[3]),
```
Comment (ods) [inline aiokafka/client.py:591]: This list is used to determine broker version. One feature is enough for each version, so this line is excessive.

**Comment by ods** [comment]
> Do you have suggestions on the types of tests you'd like to see here.

We need to proof, that this new functionality works. There must be some code with static partition assignor.

**Comment by theY4Kman** [comment]
I played around with testing this feature last year. I didn't make it very far, but I thought an important part would be the querying of consumers' group instance IDs – and for that, a v4 DescribeGroups req/res is necessary. kafka-python still does not have this record version. I started building it out in my branch:

```py
# aiokafka/protocol/admin.py

from kafka.protocol.admin import DescribeGroupsRequest_v3, DescribeGroupsResponse_v4
from kafka.protocol.api import Request, Response
from kafka.protocol.types import Array, Bytes, Int16, Int32, Schema, String


class DescribeGroupsResponse_v4(Response):
    API_KEY = 15
    API_VERSION = 4
    SCHEMA = Schema(
        ('throttle_time_ms', Int32),
        ('groups', Array(
            ('error_code', Int16),
            ('group', String('utf-8')),
            ('state', String('utf-8')),
            ('protocol_type', String('utf-8')),
            ('protocol', String('utf-8')),
            ('members', Array(
                ('member_id', String('utf-8')),
                ('group_instance_id', String('utf-8')),
                ('client_id', String('utf-8')),
                ('client_host', String('utf-8')),
                ('member_metadata', Bytes),
                ('member_assignment', Bytes)))),
            ('authorized_operations', Int32))
    )


class DescribeGroupsRequest_v4(Request):
    API_KEY = 15
    API_VERSION = 4
    RESPONSE_TYPE = DescribeGroupsResponse_v4
    SCHEMA = DescribeGroupsRequest_v3.SCHEMA
```

**Comment by ods** [comment]
We may consider removing kafka-python from dependencies, as even issues with PRs are not merged into master.

**Comment by ods** [comment]
Hello @rujutashinde, thank you for reminding. Unfortunately I don't have enough expertise to make this PR production ready. Would you like to participate?

**Comment by rujutashinde** [comment]
> Hello @rujutashinde, thank you for reminding. Unfortunately I don't have enough expertise to make this PR production ready. Would you like to participate?

I see. Unfortunately i dont have the expertise either, i am more a user of the project. But i will give it a go when i have some time cycles. Please do update the ticket if there are any new developments to this . appreciate the quick response!

**Comment by tartieret** [comment]
Hi,
I am wondering if there is anything we could do to finalize and merge this PR, as this is a super useful feature?
In case you haven't noticed and if it could be useful, the Robinhood folks implemented this in their fork of aiokafka: https://github.com/robinhood/aiokafka/pull/21/files
Let me know if any help is required

**Comment by ods** [comment]
Hi @tartieret,
Could you please provide tests for it?

**Comment by joshuaherrera** [comment]
Hello! I'd love to help get this change merged into the project, my company would benefit greatly from KIP-345 mode. You can assign this to me if you'd like.

I'll work on the requested tests. Are tests all that is preventing this from getting merged?

**Comment by ods** [comment]
Replaced with https://github.com/aio-libs/aiokafka/pull/941

**Pull request closed without merging**